addons/certificate/models/certificate.py

```diff
@@ -0,0 +1,336 @@
+        self.ensure_one()
+        cert = x509.load_pem_x509_certificate(base64.b64decode(self.with_context(bin_size=False).pem_certificate))
+        return _get_formatted_value(cert.public_bytes(serialization.Encoding.DER), formatting=formatting)
+
+    def _get_fingerprint_bytes(self, hashing_algorithm='sha256', formatting='encodebytes'):
+        self.ensure_one()
+        cert = x509.load_pem_x509_certificate(base64.b64decode(self.with_context(bin_size=False).pem_certificate))
+        if hashing_algorithm not in STR_TO_HASH:
+            raise UserError(_("Unsupported hashing algorithm. Currently supported: sha1 and sha256."))
+        return _get_formatted_value(cert.fingerprint(STR_TO_HASH[hashing_algorithm]), formatting=formatting)
+
+    def _get_signature_bytes(self, formatting='encodebytes'):
+        self.ensure_one()
+        cert = x509.load_pem_x509_certificate(base64.b64decode(self.with_context(bin_size=False).pem_certificate))
+        return _get_formatted_value(cert.signature, formatting=formatting)
+
+    def _get_public_key_numbers_bytes(self, formatting='encodebytes'):
+        self.ensure_one()
+        if self.public_key_id or self.private_key_id:
+            return (self.public_key_id or self.private_key_id)._get_public_key_numbers_bytes(formatting=formatting)
+
+        # When no keys are set to the certificate, use the self-contained public key from the content
+        return self.env['certificate.key']._get_public_key_numbers_bytes_with_key(
+            self._get_public_key_bytes(encoding='pem'),
+            formatting=formatting,
+        )
+
+    def _get_public_key_bytes(self, encoding='der', formatting='encodebytes'):
+        self.ensure_one()
+        if self.public_key_id or self.private_key_id:
+            return (self.public_key_id or self.private_key_id)._get_public_key_bytes(encoding=encoding, formatting=formatting)
+
+        # When no keys are set to the certificate, use the self-contained public key from the content
+        try:
+            cert = x509.load_pem_x509_certificate(base64.b64decode(self.with_context(bin_size=False).pem_certificate))
+            public_key = cert.public_key()
+        except ValueError:
+            raise UserError(_("The public key from the certificate could not be loaded."))
+
+        encoding = serialization.Encoding.DER if encoding == 'der' else serialization.Encoding.PEM
+        return _get_formatted_value(
+            public_key.public_bytes(
+                encoding=encoding,
+                format=serialization.PublicFormat.SubjectPublicKeyInfo
+            ),
+            formatting=formatting,
+        )
+
+    def _sign(self, message, hashing_algorithm='sha256', formatting='encodebytes'):
+        """ Return the base64 encoded signature of message. """
+        self.ensure_one()
+
+        if not self.is_valid:
+            raise UserError(_("This certificate is not valid, its validity has probably expired"))
```

Comment:
can we check if this is the case ?

Reply:
What do you mean by this comment?

Reply:
Yes something like this. Saying in the error “probably” is a bit weird. Can we make additionnal investigation on why it’s not valid?